core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -309,6 +328,109 @@ private void balance() {
+                boolean rebalanceConditionsMet = false;
+                boolean deltaAboveThreshold = false;
+                float weightWithShardAdded = Float.POSITIVE_INFINITY;
+                if (betterWeightThanCurrent) {
+                    // get the delta between the weights of the node we are checking and the node that holds the shard
+                    final float currentDelta = absDelta(nodeWeight, currentWeight);
+                    // checks if the weight delta is above a certain threshold; if it is not above a certain threshold,
+                    // then even though the node we are examining has a better weight and may make the cluster balance
+                    // more even, it doesn't make sense to execute the heavyweight operation of relocating a shard unless
+                    // the gains make it worth it, as defined by the threshold
+                    deltaAboveThreshold = lessThan(currentDelta, threshold) == false;
+                    // simulate the weight of the node if we were to relocate the shard to it
+                    weightWithShardAdded = weight.weightShardAdded(this, node, idxName);
+                    // calculate the delta of the weights of the two nodes if we were to add the shard to the
+                    // node in question and move it away from the node that currently holds it.
+                    final float proposedDelta = weightWithShardAdded - weight.weightShardRemoved(this, currentNode, idxName);
+                    rebalanceConditionsMet = deltaAboveThreshold && proposedDelta < currentDelta;
+                    // if the simulated weight delta with the shard moved away is better than the weight delta
+                    // with the shard remaining on the current node, and we are allowed to allocate to the
+                    // node in question, then allow the rebalance
+                    if (rebalanceConditionsMet && canAllocate.type().higherThan(rebalanceDecisionType)) {
+                        // rebalance to the node, only will get overwritten if the decision here is to
+                        // THROTTLE and we get a decision with YES on another node
+                        rebalanceDecisionType = canAllocate.type();
+                        assignedNodeId = node.getNodeId();
+                    }
+                }
+                nodeDecisions.put(node.getNodeId(), new NodeRebalanceDecision(
+                    rebalanceConditionsMet ? canAllocate.type() : Type.NO,
+                    canAllocate,
+                    betterWeightThanCurrent,
+                    deltaAboveThreshold,
+                    nodeWeight,
+                    weightWithShardAdded)
+                );
+            }
+
+
+            if (canRebalance.type() != Type.YES) {
+                return new RebalanceDecision(canRebalance, Type.NO, "rebalancing is not allowed", null, nodeDecisions, currentWeight);
```

Comment:
If `canRebalance.type()` returns `Type.THROTTLE`, should this pass `Type.THROTTLE` instead of hardcoding `Type.NO`?